File: sdk/swift-enhanced/Sources/FlagentEnhanced/Cache/EvaluationCache.swift
import Foundation
import FlagentClient

public struct CacheKey: Hashable {
    public let flagKey: String?
    public let flagID: Int64?
    public let entityID: String?
    public let entityType: String?
    
    public init(flagKey: String? = nil, flagID: Int64? = nil, entityID: String? = nil, entityType: String? = nil) {
        self.flagKey = flagKey
        self.flagID = flagID
        self.entityID = entityID
        self.entityType = entityType
    }
    
    func toKeyString() -> String {
        return "\(flagID?.description ?? flagKey ?? "")_\(entityID ?? "")_\(entityType ?? "")"
    }
}

private struct CachedEntry {
    let result: EvalResult
    let timestamp: Date
    
    func isExpired(ttl: TimeInterval, currentTime: Date) -> Bool {
        return currentTime.timeIntervalSince(timestamp) > ttl
    }
}

public protocol EvaluationCache {
    func get(key: CacheKey) -> EvalResult?
    func put(key: CacheKey, result: EvalResult)
    func clear()
    func evictExpired()
}

public class InMemoryEvaluationCache: EvaluationCache {
    private var cache: [String: CachedEntry] = [:]
    private let ttl: TimeInterval
    private let queue = DispatchQueue(label: "com.flagent.cache", attributes: .concurrent)
    
    public init(ttl: TimeInterval) {
        self.ttl = ttl
    }
    
    public func get(key: CacheKey) -> EvalResult? {
        return queue.sync {
            guard let entry = cache[key.toKeyString()],
                  !entry.isExpired(ttl: ttl, currentTime: Date()) else {
                cache.removeValue(forKey: key.toKeyString())
                return nil
            }
            return entry.result
        }
    }
    
    public func put(key: CacheKey, result: EvalResult) {
        queue.async(flags: .barrier) {
            self.cache[key.toKeyString()] = CachedEntry(result: result, timestamp: Date())
        }
    }
    
    public func clear() {
        queue.async(flags: .barrier) {
            self.cache.removeAll()
        }
    }
    
    public func evictExpired() {
        queue.async(flags: .barrier) {
            let currentTime = Date()
            let keysToRemove = self.cache.filter { $0.value.isExpired(ttl: self.ttl, currentTime: currentTime) }.map { $0.key }
            keysToRemove.forEach { self.cache.removeValue(forKey: $0) }
        }
    }
}

File: sdk/kotlin-enhanced/src/commonMain/kotlin/com/flagent/enhanced/cache/EvaluationCache.kt
package com.flagent.enhanced.cache

import com.flagent.client.models.EvalResult
import com.flagent.enhanced.platform.currentTimeMs
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

data class CacheKey(
    val flagKey: String?,
    val flagID: Long?,
    val entityID: String?,
    val entityType: String?
) {
    fun toKeyString(): String = "${flagID ?: flagKey}_${entityID}_${entityType}"
}

private data class CachedEntry(
    val result: EvalResult,
    val timestamp: Long
) {
    fun isExpired(ttlMs: Long, currentTime: Long): Boolean =
        (currentTime - timestamp) > ttlMs
}

interface EvaluationCache {
    suspend fun get(key: CacheKey): EvalResult?
    suspend fun put(key: CacheKey, result: EvalResult)
    suspend fun clear()
    suspend fun evictExpired()
}

class InMemoryEvaluationCache(
    private val ttlMs: Long
) : EvaluationCache {
    private val cache = mutableMapOf<String, CachedEntry>()
    private val mutex = Mutex()

    override suspend fun get(key: CacheKey): EvalResult? = mutex.withLock {
        val entry = cache[key.toKeyString()]
        if (entry != null && !entry.isExpired(ttlMs, currentTimeMs())) {
            entry.result
        } else {
            cache.remove(key.toKeyString())
            null
        }
    }

    override suspend fun put(key: CacheKey, result: EvalResult) = mutex.withLock {
        cache[key.toKeyString()] = CachedEntry(result, currentTimeMs())
    }

    override suspend fun clear() = mutex.withLock { cache.clear() }

    override suspend fun evictExpired() = mutex.withLock {
        val currentTime = currentTimeMs()
        cache.entries.filter { it.value.isExpired(ttlMs, currentTime) }.map { it.key }.forEach { cache.remove(it) }
    }
}
